book/src/proptest-derive/errors.md

```diff
@@ -0,0 +1,631 @@
+trait.
+
+## E0004
+
+This error occurs when `#[derive(Arbitrary)]` is used on an enum with no
+variants at all. For example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+enum Uninhabited { }
+```
+
+Such an enum has no values at all, so it does not make sense to provide an
+`Arbitrary` implementation for it.
+
+## E0005
+
+This error occurs if `#[derive(Arbitrary)]` is used on an enum whose variants
+are all uninhabited, using the same logic as described for [`E0003`](#e0003).
+As a result, the enum itself is totally uninhabited.
+
+Example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+enum Uninhabited {
+    V1(!),
+    V2(!, !),
+}
+```
+
+## E0006
+
+This error occurs if `#[derive(Arbitrary)]` is used on an enum where all
+inhabited variants are marked with `#[proptest(skip)]`. In other words,
+proptest is forbidden from generating any of the enum's variants, and thus the
+enum itself is unconstructable.
+
+Example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+enum MyEnum {
+    // Ordinarily, proptest would be able to generate either of these variants,
+    // but both are forbidden, so in the end proptest isn't allowed to generate
+    // anything at all.
+    #[proptest(skip)]
+    V1,
+    #[proptest(skip)]
+    V2(u32),
+    // This variant is implicitly skipped because proptest knows it is
+    // uninhabited.
+    V3(!),
+}
+```
+
+## E0007
+
+This error happens if an attribute `#[proptest(strategy = "expr")]` or
+`#[proptest(value = "expr")]` is applied to an item such as an enum or struct.
+
+Example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+#[proptest(value = "MyStruct(42)")]
+struct MyStruct(u32);
+```
+
+These attributes are only applicable to fields.
+
+## E0008
+
+This error happens if `#[proptest(skip)]` is applied to an unskippable item.
+For example, struct fields cannot be skipped because Rust requires every field
+of a struct to have a value.
+
+Example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct WidgetContainer {
+    desired_widget_count: usize,
+    #[proptest(skip)]
+    widgets: Vec<Widget>,
+}
+```
+
+In general, the appropriate way to request proptest to not generate a field
+value is to use `#[proptest(value = "expr")]` to provide a fixed value
+yourself. For example, the above code could be properly written as follows:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct WidgetContainer {
+    desired_widget_count: usize,
+    #[proptest(value = "vec![]")] // Always generate an empty widget vec
+    widgets: Vec<Widget>,
+}
+```
+
+## E0009
+
+This error happens if `#[proptest(weight = <integer>)]` is applied to an item
+where this does not make sense, such as a struct field. For example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct Foo {
+    a: u32,
+    #[proptest(weight = 42)]
+    b: u32,
+}
+```
+
+The `weight` attribute only is sensible where proptest has a choice between
+multiple items, such as enum variants. In contrast, with struct fields proptest
```

Comment:
```suggestion
multiple items, that is: enum variants. In contrast, with struct fields proptest
```

"such as" suggests that there may be more but it enum variants is the only thing literally.